src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs

```diff
@@ -331,47 +331,56 @@ internal static List<CompletionResult> MakeCommandsUnique(IEnumerable<PSObject>
-                    // in that case, we should add the module name qualification because if the module is not in the module path, calling
-                    // 'Get-Foo' directly doesn't work
-                    string completionName = keyValuePair.Key;
-                    if (!includeModulePrefix)
+                    var modulesWithCommand = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+                    var importedModules = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+                    var commandInfoArray = new CommandInfo[commandList.Count];
+                    for (int i = 0; i < commandInfoArray.Length; i++)
                     {
-                        var commandInfo = commandList[0] as CommandInfo;
-                        if (commandInfo != null && !string.IsNullOrEmpty(commandInfo.Prefix))
+                        var commandInfo = (CommandInfo)commandList[i];
+                        commandInfoArray[i] = commandInfo;
+                        if (commandInfo.CommandType == CommandTypes.Application)
                         {
-                            Diagnostics.Assert(!string.IsNullOrEmpty(commandInfo.ModuleName), "the module name should exist if commandInfo.Prefix is not an empty string");
-                            if (!ModuleCmdletBase.IsPrefixedCommand(commandInfo))
-                            {
-                                completionName = commandInfo.ModuleName + "\\" + completionName;
-                            }
+                            continue;
                         }
-                    }
 
-                    results.Add(GetCommandNameCompletionResult(completionName, commandList[0], addAmpersandIfNecessary, quote));
+                        _ = modulesWithCommand.Add(commandInfo.ModuleName);
+                        if ((commandInfo.CommandType != CommandTypes.Alias && commandInfo.CommandMetadata.CommandType is not null)
+                            || (commandInfo.CommandType == CommandTypes.Alias && commandInfo.Definition is not null))
+                        {
+                            _ = importedModules.Add(commandInfo.ModuleName);
+                        }
+                    }
 
-                    // For the other commands that are hidden, we need to disambiguate,
-                    // but put these at the end as it's less likely any of the hidden
-                    // commands are desired.  If we can't add anything to disambiguate,
-                    // then we'll skip adding a completion result.
-                    for (int index = 1; index < commandList.Count; index++)
+                    int moduleCount = modulesWithCommand.Count;
+                    modulesWithCommand.Clear();
+                    int index;
+                    if (commandInfoArray[0].CommandType == CommandTypes.Application
+                        || importedModules.Count == 1
+                        || moduleCount < 2)
```

Comment:
Can you please explain this check too? Especially `moduleCount < 2`.

Reply:
It's to check if the short name for the command can be used. If the first element is an application then we know there's no conflicting commands/aliases (because of the command precedence in PS).
If there's just 1 module imported then the short name refers to that module (and it will be the first element in the list) and of course if there's less than 2 unique modules exporting that command then we can use the short name because it can only refer to that module.